### PR TITLE
Unwrap substitutions both before _and_ after potential simplification

### LINKED ISSUE #31804: Generic wrapper interface no longer assignable
**TypeScript Version:**  3.6.0-dev.20190606

**Search Terms:**
I didn't know how to search for this issue since it's a bit complex

**Code**

```ts
type AnyFunction = (...args: any[]) => any;
type Params<T> = Parameters<Extract<T, AnyFunction>>;

interface Wrapper<T> {
	call<K extends keyof T>(event: K, ...args: Params<T[K]>): void;
}

interface AWrapped {
	foo(): void;
}

class A {
	foo: Wrapper<AWrapped>;
}

interface BWrapped extends AWrapped {
	bar(): void;
}

class B extends A {
	foo: Wrapper<BWrapped>;
}
```

**Expected behavior:**
No errors, as it worked in pre 3.5.0

**Actual behavior:**
> Property 'foo' in type 'B' is not assignable to the same property in base type 'A'.
>   Type 'Wrapper<BWrapped>' is not assignable to type 'Wrapper<AWrapped>'.
>     Property 'bar' is missing in type 'AWrapped' but required in type 'BWrapped'

**Why? What was this used for?**
Our project uses a setup like this for strongly typed events; we define the events in interfaces, and then classes that emit the events store the generic event emitter instances on them. (The actual implementation is a lot more complex, this is a stripped down version) 

Previously, it worked to simply override the property's type, but now that doesn't work anymore, and I can't find a workaround. I think this is might prevent our team from updating because of how much this breaks, at least for a time... The amount that has to be rewritten to undo the dependency on this functionality is not negligible.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #31804

The prophecized bug from when we added conditional simplification has come to pass, thus we must decide: Do we continue being naive and just apply a second unwrap step (as I have added here), or do we complicate the unwrapping/simplifying a bit more to guarantee that we always get the final maximally unwrapped/simplified type (as was originally in the conditional simplification PR).

### CODE REVIEW

**Comment by weswigham** [comment]
reping @ahejlsberg ❤️

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -12529,6 +12529,14 @@ namespace ts {
...
                     target = getSimplifiedType(target, /*writing*/ true);
                 }
+                if (source.flags & TypeFlags.Substitution) {
+                    source = (<SubstitutionType>source).substitute;
+                }
+                if (target.flags & TypeFlags.Substitution) {
+                    target = (<SubstitutionType>target).typeVariable;
+                }
+                Debug.assert(!(source.flags & TypeFlags.Substitution), "Source type was a substitution - substitutes are unhandled in relationship checking");
+                Debug.assert(!(target.flags & TypeFlags.Substitution), "Target type was a substitution - substitutes are unhandled in relationship checking");
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Perhaps this instead:

```ts
while (true) {
    const s = source.flags & TypeFlags.Substitution ? source = (<SubstitutionType>source).substitute :
        source.flags & TypeFlags.Simplifiable ? getSimplifiedType(source, /*writing*/ false) :
        source;
    const t = target.flags & TypeFlags.Substitution ? (<SubstitutionType>target).typeVariable :
        target.flags & TypeFlags.Simplifiable ? getSimplifiedType(target, /*writing*/ true) :
        target;
    if (s === source && t === target) break;
        source = s;
        target = t;
    }
}
```

**Comment by weswigham** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt

File: src/compiler/checker.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/checker.ts]: @ahejlsberg done - let's see if anything stacks out on DT~

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -12776,17 +12776,16 @@ namespace ts {
...
-                if (target.flags & TypeFlags.Substitution) {
-                    target = (<SubstitutionType>target).typeVariable;
-                }
-                if (source.flags & TypeFlags.Simplifiable) {
-                    source = getSimplifiedType(source, /*writing*/ false);
-                }
-                if (target.flags & TypeFlags.Simplifiable) {
-                    target = getSimplifiedType(target, /*writing*/ true);
+                while (true) {
+                    const s = source.flags & TypeFlags.Substitution ? source = (<SubstitutionType>source).substitute :
```
Comment (jack-williams) [inline src/compiler/checker.ts]: Would it ever be worth updating `substitute` in-place to collapse multiple links down to one?

**Comment by weswigham** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

File: src/compiler/checker.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/checker.ts]: I don't think we can - the associated `typeVariable` of nested `SubstitutionType`s can (and most often does) differ. ie, a `T` substituted with `T & U` where `U` is substituted with `U & string` - while here we want to fully flatten, in other situations I think we want to keep the knowledge of what variables expanded to what.

Comment (jack-williams) [reply src/compiler/checker.ts]: Right; I forgot that these are more like narrowings rather than unification substitutions.

**Comment by weswigham** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by weswigham** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by weswigham** [comment]
RWC is 👍 (no diff with `master`)
User is 👍  (same `vue-next` CI only (?) failure as `master`)
DT is ✅ (no failures)
Perf is ongoing but should be OK  (it shouldn't have changed since the last run; I doubt this plays poorly with something recent in `master`)

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -15130,8 +15130,18 @@ namespace ts {
...
+                    const s = getNormalizedType(source, /*writing*/ false);
+                    if (s === source) break;
+                    source = s;
+                } while (true);
+                let target = originalTarget;
+                do {
+                    const t = getNormalizedType(target, /*writing*/ true);
+                    if (t === target) break;
+                    target = t;
+                } while (true);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Put the loop inside `getNormalizedType`. It is only called from here.

**Pull request merged**